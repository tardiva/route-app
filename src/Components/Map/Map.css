.map {
    width: 100%;
    height: 100vh;
}

.map__warning {
    display: flex;
    justify-content: center;
    align-items: center;
    height: 100%;
    color: #e83232;
}