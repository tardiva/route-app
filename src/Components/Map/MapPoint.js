export default class MapPoint {
  constructor(mapProvider, point) {
    this.id = point.id;
    this.marker = new mapProvider.maps.Marker({
      position: new mapProvider.maps.LatLng(point.coord.lat, point.coord.lng),
      title: point.name,
      draggable: true
    });
  }

  setEventHandler(eventName, handler) {
    this.marker.addListener(eventName, () => {
      handler(this);
    });
  }

  getPosition() {
    return this.marker.getPosition();
  }

  setMap(map) {
    this.marker.setMap(map);
  }
}