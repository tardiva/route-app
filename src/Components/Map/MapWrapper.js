import makeAsyncScriptLoader from 'react-async-script';
import Map from './Map';

const key = 'AIzaSyCQfJgnHnO6LlY_DxE-VaYEB_qWV-W9O7c';
const url = `https://maps.googleapis.com/maps/api/js?key=${key}`;
const globalName = 'google';

export default makeAsyncScriptLoader(url, {
  globalName: globalName
})(Map);