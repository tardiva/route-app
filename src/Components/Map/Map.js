import React, { Component } from 'react';
import MapPoint from './MapPoint';
import './Map.css';

class Map extends Component {

  constructor(props) {
    super(props);
    this.map = null;
    this.route = null;
    this.infowindow = null;
    this.mapPoints = [];
    this.handleMapChange = this.handleMapChange.bind(this);
    this.handlePointInfo = this.handlePointInfo.bind(this);
    this.handlePointMove = this.handlePointMove.bind(this);
    this.handlePointHasMoved = this.handlePointHasMoved.bind(this);
  }

  componentDidUpdate(prevProps) {
    const { google } = this.props;
    if (!prevProps.google && google) {
      this.initMap();
      this.initRoute();
      this.initInfowindow();
      this.map.addListener('center_changed', this.handleMapChange);
    }
    if (prevProps.points !== this.props.points) {
      this.updateRoute(prevProps.points, this.props.points)
    }
  }

  render() {
    return (
      <div ref="map" className="map">
        {this.props.warning ? (<div className="map__warning"><span>Что-то пошло не так. Не удается загрузить карту.</span></div>) : ''}
      </div>
    );
  }

  initMap() {
    const { google } = this.props;
    this.map = new google.maps.Map(this.refs.map, {
      zoom: 12,
      center: new google.maps.LatLng(this.props.center.lat, this.props.center.lng),
      mapTypeControl: false,
      streetViewControl: false
    });
  }

  initRoute() {
    this.route = new this.props.google.maps.Polyline({
      geodesic: false,
      strokeColor: '#FF0000',
      strokeOpacity: 1.0,
      strokeWeight: 2
    });
    this.route.setMap(this.map);
  }

  initInfowindow() {
    this.infowindow = new this.props.google.maps.InfoWindow({});
  }

  handleMapChange() {
    this.props.handleMapChange(this.transformCoordinates(this.map.getCenter()));
  }

  updateRoute(prevPoints, newPoints) {

    // remove outdated points
    prevPoints.forEach((point) => {
      if (!newPoints.find((newPoint) => newPoint.id === point.id)) {
        this.deletePoint(point.id);
      }
    });

    // add new points
    newPoints.forEach((point, index) => {
      const prevIndex = prevPoints.findIndex((prevPoint) => prevPoint.id === point.id);
      if (prevIndex < 0) {
        this.addPoint(point, index);
      }
    });

    // reorder points
    this.mapPoints.forEach((mapPoint, index) => {
      const id = mapPoint.id;
      if (id !== this.props.points[index].id) {
        const prevIndex = this.mapPoints.findIndex((point) => id === point.id);
        const nextIndex = this.props.points.findIndex((point) => id === point.id);
        this.reorderPoint(prevIndex, nextIndex);
      }
    });
  }

  addPoint(point, index) {
    const mapPoint = new MapPoint(this.props.google, point);

    mapPoint.setEventHandler('click', this.handlePointInfo);
    mapPoint.setEventHandler('drag', this.handlePointMove);
    mapPoint.setEventHandler('dragend', this.handlePointHasMoved);

    mapPoint.setMap(this.map);
    this.mapPoints.splice(index, 0, mapPoint);
    this.route.getPath().insertAt(index, mapPoint.getPosition());
  }

  deletePoint(id) {
    const index = this.mapPoints.findIndex(mapPoint => mapPoint.id === id);
    this.mapPoints[index].setMap(null);
    this.mapPoints.splice(index, 1);
    this.route.getPath().removeAt(index);
  }

  reorderPoint(prevIndex, nextIndex) {
    this.mapPoints.splice(nextIndex, 0, this.mapPoints.splice(prevIndex, 1)[0]);
    const path = this.route.getPath();
    const point = path.getAt(prevIndex);
    path.removeAt(prevIndex);
    path.insertAt(nextIndex, point);
  }

  handlePointInfo(mapPoint) {
    const marker = mapPoint.marker;
    this.infowindow.setContent(`<div class="map__infowindow">${marker.title}</div>`);
    this.infowindow.open(this.map, marker);
  }

  handlePointMove(mapPoint) {
    this.route.getPath().setAt(this.getPointIndexById(mapPoint.id), mapPoint.getPosition());
  }

  handlePointHasMoved(mapPoint) {
    this.props.handlePointMove(this.getPointIndexById(mapPoint.id), this.transformCoordinates(mapPoint.getPosition()));
  }

  getPointIndexById(id) {
    return this.mapPoints.findIndex(point => point.id === id);
  }

  transformCoordinates(googleLatLng) {
    return {lat: googleLatLng.lat(), lng: googleLatLng.lng()}
  }
}

export default Map;