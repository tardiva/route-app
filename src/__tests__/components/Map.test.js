import React from 'react';
import { shallow } from 'enzyme';
import Map from '../../Components/Map/Map';
import points from '../../Fixtures/points';
import google from '../../__mocks__/google.mock';

it('should map render correctly', () => {
  const wrapper = shallow(<Map center={{lat: 0, lng: 0}} points={[]}/>);
  expect(wrapper).toMatchSnapshot();
});

it('should show error message when map is unavailable', () => {
  const wrapper = shallow(<Map center={{lat: 0, lng: 0}} points={[]} warning={true}/>);
  expect(wrapper).toMatchSnapshot();
});

it('should init map correctly', () => {
  const wrapper = shallow(<Map center={{lat: 0, lng: 0}} points={[]}/>);
  const component = wrapper.instance();
  expect(component.map).toBeFalsy();
  expect(component.route).toBeFalsy();
  expect(component.infowindow).toBeFalsy();
  wrapper.setProps({google: google});
  expect(google.maps.Map).toHaveBeenCalledTimes(1);
  expect(component.map).toBeTruthy();
  expect(google.maps.Polyline).toHaveBeenCalledTimes(1);
  expect(component.route).toBeTruthy();
  expect(google.maps.InfoWindow).toHaveBeenCalledTimes(1);
  expect(component.infowindow).toBeTruthy();
});

it('should handle map center change', () => {
  const onMapChangeSpy = jest.fn();
  const wrapper = shallow(<Map center={{lat: 0, lng: 0}} points={[]} handleMapChange={onMapChangeSpy}/>);
  wrapper.setProps({google: google});
  const component = wrapper.instance();
  const map = component.map;
  map.trigger('center_changed');
  expect(onMapChangeSpy).toHaveBeenCalledTimes(1);
  expect(map.getCenter).toHaveBeenCalledTimes(1);
});

it('should add map points', () => {
  const wrapper = shallow(<Map center={{lat: 0, lng: 0}} points={[]}/>);
  const component = wrapper.instance();
  wrapper.setProps({google: google, points: [points[0]]});
  expect(component.mapPoints).toHaveLength(1);
});

it('should delete map points', () => {
  const wrapper = shallow(<Map center={{lat: 0, lng: 0}} points={[]}/>);
  const component = wrapper.instance();
  wrapper.setProps({google: google, points: [points[0], points[1]]});
  expect(component.mapPoints).toHaveLength(2);
  wrapper.setProps({google: google, points: [points[0]]});
  expect(component.mapPoints).toHaveLength(1);
});

it('should reorder map points', () => {
  const wrapper = shallow(<Map center={{lat: 0, lng: 0}} points={[]}/>);
  const component = wrapper.instance();
  wrapper.setProps({google: google, points: [points[0], points[1]]});
  expect(component.mapPoints).toHaveLength(2);
  wrapper.setProps({google: google, points: [points[1], points[0]]});
  expect(component.mapPoints).toHaveLength(2);
});

it('should open infowindow when clicking on map point', () => {
  const wrapper = shallow(<Map center={{lat: 0, lng: 0}} points={[]}/>);
  const component = wrapper.instance();
  wrapper.setProps({google: google, points: points});
  component.mapPoints[0].marker.trigger('click');
  expect(component.infowindow.open).toHaveBeenCalledTimes(1);
});

it('should call handlePointMove prop when dragging map point', () => {
  const onDragEndSpy = jest.fn();
  const wrapper = shallow(<Map center={{lat: 0, lng: 0}} points={[]}/>);
  const component = wrapper.instance();
  wrapper.setProps({google: google, points: points, handlePointMove: onDragEndSpy});
  component.mapPoints[0].marker.trigger('drag');
  component.mapPoints[0].marker.trigger('dragend');
  expect(onDragEndSpy).toHaveBeenCalledTimes(1);
});






