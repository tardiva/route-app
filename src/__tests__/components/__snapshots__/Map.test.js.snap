// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should map render correctly 1`] = `
<div
  className="map"
/>
`;

exports[`should show error message when map is unavailable 1`] = `
<div
  className="map"
>
  <div
    className="map__warning"
  >
    <span>
      Что-то пошло не так. Не удается загрузить карту.
    </span>
  </div>
</div>
`;
